File: angular-ecommerce-book-store/ecommerce-book-store/src/app/common/product.ts
export interface Product {
    id: number;
    category: Category;
    author: Author;
    language: Language;
    isbn: string;
    name: string;
    description: string;
    unitPrice: number;
    image: string | null;
    active: boolean;
    unitsInStock: number;
    dateCreated: Date;
    lastUpdated: Date;
    numOfPages: number;
    yearOfPublication: number;
  }

export interface ProductCreate{

    id: number;
    categoryId: number;
    authorId: number;
    languageId: number;
    isbn: string;
    name: string;
    description: string;
    unitPrice: number;
    image: string | null;
    active: boolean;
    unitsInStock: number;
    dateCreated: Date;
    lastUpdated: Date;
    numOfPages: number;
    yearOfPublication: number;
}

export class ProductEdit{

  id: number;
  categoryId: number;
  authorId: number;
  languageId: number;
  isbn: string;
  name: string;
  description: string;
  unitPrice: number;
  image: string | null;
  active: boolean;
  unitsInStock: number;
  dateCreated: Date;
  lastUpdated: Date;
  numOfPages: number;
  yearOfPublication: number;

  constructor(product:Product){
      this.id = product.id;
      this.categoryId = product.category.id;
      this.authorId = product.author.id;
      this.languageId = product.language.id;
      this.isbn = product.isbn;
      this.name = product.name;
      this.description = product.description;
      this.unitPrice = product.unitPrice;
      this.image = product.image;
      this.active = product.active;
      this.unitsInStock = product.unitsInStock;
      this.dateCreated = product.dateCreated;
      this.lastUpdated = product.lastUpdated;
      this.numOfPages =product.numOfPages;
      this.yearOfPublication = product.yearOfPublication;
  }
}

  export interface Category {
    id: number;
    categoryName: string;
  }
  
  export interface Author {
    id: number;
    firstName: string;
    lastName: string;
  }
  
  export interface Language {
    id: number;
    name: string;
    designation: string;
  }

File: spring-boot-ecommerce-book-store/src/main/java/com/example/springbootecommercebookstore/entity/Product.java
package com.example.springbootecommercebookstore.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.sql.Blob;
import java.util.Date;

@Entity
@Table(name = "product")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Product {


    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private ProductCategory category;

    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    private Author author;

    @ManyToOne
    @JoinColumn(name = "language_id", nullable = false)
    private Language language;

    @Column(name = "isbn")
    private String isbn;

    @Column(name = "name")
    private String name;

    @Column(name = "description")
    private String description;

    @Column(name = "unit_price")
    private BigDecimal unitPrice;

    @Lob
    @Column(name = "image")
    private Blob image;

    @Column(name = "active")
    private boolean active;

    @Column(name = "units_in_stock")
    private int unitsInStock;

    @Column(name = "date_created")
    @CreationTimestamp
    private Date dateCreated;

    @Column(name = "last_updated")
    @UpdateTimestamp
    private Date lastUpdated;

    @Column(name = "num_of_pages")
    private int numOfPages;

    @Column(name = "year_of_publication")
    private int yearOfPublication;
}
